### ISSUE TITLE
Same program much slower in 0.4.0 than in 0.3.0

### ISSUE THREAD

**Issue by HMEIatJHU** [pytorch/pytorch#6694]
- PyTorch
- OS: Ubuntu 14.04 LTS
- PyTorch version: 0.4.0.a0+a589180 and 0.3.1.post2
- Python version: 3.6

My program runs much slower (~50%) in 0.4.0 than in 0.3.0 with the same set of parameters. 
What could be the most important reasons? 

Is there some general tips that I should follow, when I adapt from 0.3.0 to 0.4.0, to make my program equally fast? 

Thanks very much in advance!!!

**Comment by SsnL** [comment]
Can you provide a repro script?

**Comment by HMEIatJHU** [comment]
@SsnL i think i better submit this to forum instead of issues... close it

**Comment by SsnL** [comment]
well if it is a bug, then it should be here.

**Comment by HMEIatJHU** [comment]
@SsnL Thanks very much for prompt response. My program is a bit large and I guess I need to profile it in details before providing a repro script. But in general, I suspect if the merging of data and Variable causes this problem. My program uses a trained model to give my output, and then use the output to do a lot of complex computation. In my code, I always use output.data instead of output, so the graph is not growing for further computation. But 0.4.0 does not distinguish Variable and data, so while output is used in following functions, the graph may keep expanding and some more history is stored. That may be the issue? Does it make sense to you? I tried to use .detach() wherever I believe it should be, but it did not solve the problem. I suspect that my habit of speeding code up in 0.3 does not apply to 0.4...

**Comment by SsnL** [comment]
On 0.4 .data always detaches and stops tracking history.

**Comment by HMEIatJHU** [comment]
@fmassa thanks for the insight! yes, I do have a lot of complicated tensor indexing and some are advanced. any tips for this?

**Comment by fmassa** [comment]
Depending on the I sexing you have, using `index_select` will definitely be faster if you only index on a single dimension

**Comment by HMEIatJHU** [comment]
@fmassa Thanks! I extensively used things like: 

1) 
ind0 = a > 0.0 & b < 0.0 # a is a high-dim tensor
ind1 = a < 0.0 & b > 0.0
a[ind0 | ind1] = 0.0

2) 
a[ind0 & ind1] = b[ind0 & ind1] + c[ind0 & ind1]
# ind0, ind1 are ByteTensor
# a, b, c are FloatTensor 
# they are all updated in complex ways 

3) 
b = a[range, inds, :] 
# a is a 3-dim tensor 
# range is a 1-dim tensor of [0,1,2,3,...]
# inds is a 1-dim tensor of ints (as positions)
# so b[i,:] = a[i,inds[i], :] 

4) 
a[range, step] = b 
# a is a 2-dim tensor and b is 1-dim tensor 
# range is 1-dim tensor of [0,1,2,3,...]
# step is 1-dim tensor of ints
# so a[ i,step[i] ] = b[i] 

Can you suggest some good ways of doing these things?

**Comment by fmassa** [comment]
Sure, here are a few options:

1. Use `masked_fill_`
2. Use a combination of `masked_select` and `masked_copy_`

For 3 and 4, I think there is no significant slowdown for this case compared to 0.3, so you can still use it.

And we are aware of those slowdowns, they are being tracked in https://github.com/pytorch/pytorch/issues/5673#issuecomment-372980255 

Thanks!

**Comment by fmassa** [comment]
@HMEIatJHU could you try installing a more recent version from master and see if your slowdowns are still as severe? Since https://github.com/pytorch/pytorch/pull/6793 several indexing cases were optimized

**Comment by fmassa** [comment]
More specifically, your case 1 should be more optimized. We have not modified parts of case 2 though

**Comment by HMEIatJHU** [comment]
@fmassa Thanks very much! 
Yes, I tried with current master today and it is good. And I will start using masked_fill/select/copy at right places.

**Comment by fmassa** [comment]
@HMEIatJHU after the #6793, there is no more need to try to use `masked_fill` etc, and you can directly use indexing, as before.

The only case where you could see some speedup by using those legacy functions is the case of `index_copy_` and `masked_scatter_`. So for example, in your case 2:
```python
ind = ind0 & ind1
a[ind] = b[ind] + c[ind]
```
You could have eventually better performances if you did
```python
ind = ind0 & ind1
values = b[ind] + c[ind]
a.masked_scatter_(ind, values)
```

Could you have a look and verify if that changes a lot the runtime?

**Comment by HMEIatJHU** [comment]
@fmassa Thanks!  

I tried a simple script as below on 3 different versions of PyTorch on Ubuntu 1404: 
```python 
import torch
import timeit

torch.manual_seed(123456)

dim = 1000
number = 1000

def use_index(a, b, c):

    ind0 = (a > 0.0) & (b < 1.0)
    ind1 = (a < 0.0) & (b > 1.0)
    ind = ind0 & ind1
    c[ind] = a[ind] + b[ind]

def use_legacy_function(a, b, c):

    ind0 = (a > 0.0) & (b < 1.0)
    ind1 = (a < 0.0) & (b > 1.0)
    ind = ind0 & ind1
    values = a[ind] + b[ind]
    c.masked_scatter_(ind, values)

def wrapper(func, *args, **kwargs):
    def wrapper():
        return func(*args, **kwargs)
    return wrapper

for device in [torch, torch.cuda]:

    print("time it on {}".format('gpu' if device==torch.cuda else 'cpu'))

    a = device.FloatTensor(dim, dim).normal_(0.0, 1.0)
    b = device.FloatTensor(dim, dim).exponential_(1.0)
    c = device.FloatTensor(dim, dim)

    wrapped = wrapper(use_index, a, b, c)
    print("use indexing, time is : {}".format(
        timeit.timeit(wrapped, number=number)) )

    wrapped = wrapper(use_legacy_function, a, b, c)
    print("use legacy functions, time is : {}".format(
        timeit.timeit(wrapped, number=number)) )
```

Here are results with different versions: 

0.3.1.post2
```python
time it on cpu
use indexing, time is : 8.527695072000142
use legacy functions, time is : 8.785706020000134
time it on gpu
use indexing, time is : 1.073753952000061
use legacy functions, time is : 1.0569869740002105
```

0.4.0a0+a589180
```python 
time it on cpu
use indexing, time is : 7.870821144999923
use legacy functions, time is : 7.257624652000004
time it on gpu
use indexing, time is : 1.2752332359999627
use legacy functions, time is : 1.2200424609999345
```

0.4.0a0+a589180
```python 
time it on cpu
use indexing, time is : 7.808169143999976
use legacy functions, time is : 7.182136361000062
time it on gpu
use indexing, time is : 1.1385018550000723
use legacy functions, time is : 1.0596150669999815
```

**Comment by fmassa** [comment]
Thanks for the tests! The results you showed look reasonable to me. There is still some room for improvement in some indexing cases but it's already much better than before.
And I'm curious, is your script now still much slower than before (using latest master compared to the version you used when opening this issue)?

**Comment by HMEIatJHU** [comment]
@fmassa not any more. it seems roughly at the same speed (not 100% sure cuz of randomness in the program). Thanks very much!

**Comment by fmassa** [comment]
In the end this fix was not included in 0.4.
Can you check if it's due to the indexing issues I pointed? Specifically, if changing in some cases to use `index_select` the makes your code run faster?

**Comment by amoussawi** [comment]
well I suspect that it's due to indexing, but not sure. I don't need `index_select` in the function that I think has slowed down, I'm just looping through a 1-D tensor, something like this
```python
indices = sparse_tensor._indices()
columns = indices[1] # the size can be in thousands

for col in range(columns.size(0)):
  # reading columns[col]
```

**Comment by amoussawi** [comment]
sorry, and in addition to the above
```python
indices = sparse_tensor._indices()
indices_2 = torch.LongTensor(indices.size())
# reading indices[0, :] and indices[1, :]
# writing indices_2[0, :] and indices_2[1, :]
```

**Comment by fmassa** [comment]
Accessing individual elements of a tensor is slower in PyTorch 0.4, but I believe only if the returned element is a scalar, not a tensor. I.e., indexing a 1d tensor is slower

**Comment by amoussawi** [comment]
Yes the returned element is scalar in my case. 
Is this going to be fixed in some new release soon?

**Comment by fmassa** [comment]
This will be fixed at some point, but I'm not sure when. It's a high-priority issue and is being tracked in https://github.com/pytorch/pytorch/issues/5388

**Comment by amoussawi** [comment]
Thank you! do you suggest any workaround?

**Comment by fmassa** [comment]
not sure what you are doing actually, but if you don't need to backprop through the elements and just want to query individual elements from a tensor, maybe converting the tensor to a list `.tolist()` could help.

**Comment by amoussawi** [comment]
So yes it worked. Thanks! 
it is way faster, but still slightly slower than when I run it in 0.3.
btw, I have the same function written in c++ (extension) and python, it looks like the issue of indexing is not that apparent in the c++ version as in the python version, but the code sped up when I indexed the tensor internal data `.data_ptr()` rather than the tensor itself.

**Comment by fmassa** [comment]
The main reason of the slowdown is that now when we index a 1d tensor we return a scalar, and not a python number. But creating a Tensor struct with a single element brings an additional overhead, because we need to allocate a bunch of things (like sizes, strides, etc).

**Comment by amoussawi** [comment]
is it done this way so that the indexing is backprop-able ?
